proxy/shadowsocks2022/encoding.go

```diff
@@ -0,0 +1,257 @@
+package shadowsocks2022
+
+import (
+	"bytes"
+	"crypto/cipher"
+	cryptoRand "crypto/rand"
+	"io"
+	"math/rand"
+	"time"
+
+	"github.com/lunixbochs/struc"
+
+	"github.com/v2fly/v2ray-core/v5/common/buf"
+	"github.com/v2fly/v2ray-core/v5/common/crypto"
+	"github.com/v2fly/v2ray-core/v5/common/net"
+	"github.com/v2fly/v2ray-core/v5/common/protocol"
+)
+
+type TCPRequest struct {
+	keyDerivation KeyDerivation
+	method        Method
+
+	c2sSalt  RequestSalt
+	c2sNonce crypto.BytesGenerator
+	c2sAEAD  cipher.AEAD
+
+	s2cSalt  RequestSalt
+	s2cNonce crypto.BytesGenerator
+	s2cAEAD  cipher.AEAD
+
+	s2cSaltAssert         RequestSalt
+	s2cInitialPayloadSize int
+}
+
+func (t *TCPRequest) EncodeTCPRequestHeader(effectivePsk []byte,
+	eih [][]byte, address DestinationAddress, destPort int, initialPayload []byte, out *buf.Buffer,
+) error {
+	requestSalt := newRequestSaltWithLength(t.method.GetSessionSubKeyAndSaltLength())
+	{
+		err := requestSalt.FillAllFrom(cryptoRand.Reader)
+		if err != nil {
+			return newError("failed to fill salt").Base(err)
+		}
+	}
+	t.c2sSalt = requestSalt
+	sessionKey := make([]byte, t.method.GetSessionSubKeyAndSaltLength())
+	{
+		err := t.keyDerivation.GetSessionSubKey(effectivePsk, requestSalt.Bytes(), sessionKey)
+		if err != nil {
+			return newError("failed to get session sub key").Base(err)
+		}
+	}
+
+	aead, err := t.method.GetStreamAEAD(sessionKey)
+	if err != nil {
+		return newError("failed to get stream AEAD").Base(err)
+	}
+	t.c2sAEAD = aead
+	paddingLength := TCPMinPaddingLength
+	if initialPayload == nil {
+		initialPayload = []byte{}
+		paddingLength += rand.Intn(TCPMaxPaddingLength) // TODO INSECURE RANDOM USED
```

Comment:
In section 3.1.3 of the spec:

> When making a request header, if payload is not available, add non-zero random length padding.

It's fine to use simple RNGs. But sending a request with no initial payload and no padding will get yourself rejected by spec-compliant servers.

```suggestion
		paddingLength += 1 + rand.Intn(TCPMaxPaddingLength)
```